packages/core/src/render3/features/host_directives_feature.ts

```
@@ -69,7 +70,7 @@ export function ɵɵHostDirectivesFeature(
  * @param matches Directives resolved through selector matching.
  */
 function resolveHostDirectives(matches: DirectiveDef<unknown>[]): HostDirectiveResolution {
-  const allDirectiveDefs: DirectiveDef<unknown>[] = [];
+  const allDirectiveDefs = new Set<DirectiveDef<unknown>>();
```

The reason why we haven't done this ourselves, but rather we throw the error, is that directives can influence the values during DI, as well as the order in which host bindings are applied. There isn't really a good way to know if we should pick the first match or one of the subsequent ones.

---

If you had something like this, assuming that `dirFoo` is a directive that is applied twice as a host directive:

```
<div dirFoo dirBar dirBaz dirFoo></div>
```

The host bindings on the `div` will execute based on the matching order of the directives. This means that depending on if you decide to take the first instance of `dirFoo` or the second one, the result can be different. E.g. if both `dirFoo` and `dirBar` had a host binding of `[class.active]`, taking the first instance of `dirFoo` will allow `dirBar` to override it, whereas if you take the second instance, `dirFoo` will override `dirBar`.

---

`<div dirFoo dirFoo></div>` still produces NG0309

---

> There isn't really a good way to know if we should pick the first match or one of the subsequent ones.

While this might be true it is absolutely clear that picking a first match is way better than throwing in runtime.

---

> I'd be hesitant to add this automatically, but we could potentially allow the consumer of the host directive to opt into the de-duplication.

How do you see it?

```
@Directive({
    deduplicationStrategy: 'none' | 'first' | 'last';
})
class DirFoo {}
```

When 'none' passed (by default), then current behavior keeps in.

When 'first' or 'last' passed, then we select first or last matched directive.

---

I like your variant. I'd like to take feedback from @waterplea, @ashley-hunter: Do we need to specify which directive we should take (first or last)? Or it is enough to take first always (if `deduplication: true` passed)?

---

> I like your variant

But, how we can handle this situation, then?:

```typescript
@Directive()
class DirFoo {}

@Directive({
   hostDirectives: [DirFoo],
})
class DirBar {}

@Directive({
   hostDirectives: [{directive: DirFoo, deduplicate: true}],
})
class DirBaz {}

@Component({
   hostDirectives: [DirBar, DirBaz],
})
class Cmp {}
```

---

> one instance of the host directive had the duplication flag and another (of the same directive) did not

It can be two, or more. Any possible combination.

> Or perhaps, it would be a requirement that all instances must be marked with the deduplicate flag or it will throw.

Then I think it is reasonable to add such flag to `@Directive` decorator instead. It applies to all instances.

---

I'd say the main issue is predictability. If we silently de-duplicate the directives, it won't be obvious to users what the execution order is, unless they're familiar with the internals. I agree that the current situation isn't great, but we need to be careful not to introduce something that will be even harder to fix. E.g. we can always remove the runtime assertion, but adding something in later will be way more breaking.

---

We were discussing this within the team, but it wasn't actually clear in which specific cases it comes up. Can you point to some examples where this is a problem?